### ISSUE TITLE
Main gesture doesn't work with cubase 13

### ISSUE THREAD

**Issue by eremsoft** [pnoqable/PinchBar#2]

**Comment by p33k** [comment]
+1

**Comment by pnoqable** [comment]
Hey guys,

tldr: No licence to test right now, so this issue depends on your help. Could you plz clarify:

1. Is preset "Cubase" still automatically activated when using Cubase 13 (see picture below)?
2. Does the issue persist after changing preset to "None" and then back to "Cubase"?
3. Are vertical and waveform zoom (pinch while pressing CMD / OPT) broken in Cubase 13, too?
4. Did any zoom conventions change with Cubase 13? I.e.:
4.1. Does normal mouse zoom (not pinch, but scroll while pressing CMD) still work?
4.1. Does pressing G/H still work for vertical zoom?
4.3. Does pressing OPT+G/H still work for waveform zoom?
5. If somebody uses such: Do other automation tools (like Automator, BetterTouchTool, ...) still work with Cubase 13?

Just to clarify, this is what PinchBar should look like when (any) Cubase is active:

<img width="65%" src="https://github.com/pnoqable/PinchBar/assets/38157207/694f047a-80d5-4f1f-8821-d85cefd728bc">

In the last months I extended PinchBar for many other applications and tested it with other people for many different use cases with a lot of different apps. Besides Cubase 13 there are absolutely no signs of incompatibility with any other application I know of. It seems quite odd to me why PinchBar just stops working with this new version of Cubase whereas always working flawlessly with all of its predecessors...

Thank you for your help!

<details>
<summary><u>Long story</u></summary>

I know of this issue since beta testing of Cubase 13 started. I requested access to the beta programme, but unfortunately my request was not accepted. Also, there are legal restrictions preventing beta testers to give outsiders detailed information about the beta and its changes in compability and whatsoever. So I haven't been able to work on this issue, yet. Furthermore, I had absolutely no success in getting in touch with anybody official at Steinberg about PinchBar, which is a real bummer to me.

Following was just me being paranoid:

~~I know that I'm just a single nerd on Steinbergs platform and they got way bigger business to manage than what I contribute to their community. But now I wonder: What if they know about, but really dislike this backwards compatible, free-to-use feature advertised on their platform? Maybe they don't want no free pinch-to-zoom feature and rather keep it closed as a (small) selling feature for the next major release of Cubase or so. In that case, i.e. if I knew that for certain, I wouldn't want to interfere with them. After all, it's their product and I like it. And I also don't have the burden of selling it again and again, year after year, since freaking 34 years! I really do appreciate Steinberg as a strong and robust tech company here in Northern Germany!~~

~~It's just that - if something like this really happened - if they read my request to talk to them - but they dismissed it and rather implemented some kind of countermeasure against PinchBar - what would that imply about their view on their community? In this hypothetical (!) situation, wouldn't it have been way smarter to make PinchBar obsolete by implementing propper trackpad support rather than implementing countermeasures and with that blocking significant community efforts, re-opening a somewhat half-closed feature request? I really hope this is not what happened here...~~

So, that's my reasoning:

Personally, I don't have real need for Cubase 13 right now and I already invested $100 to distribute PinchBar to you guys with the signature of an official Apple Developer Certificate, which by now expired and needs to be renewed for the next release of PinchBar, adding another $100 to the bill. ATM I'm questioning wheather it would be wise to invest 80€ for the upgrade to Cubase 13, just to invest more time into PinchBar, analysing the current situation - and who knows - maybe only to find out that Steinberg really implemented countermeasures to prevent apps like PinchBar from interacting with their UI.

However, this wouldn't be the end of the rabbit hole as I have some experience with lower level kernel space event handling in macOS. I guess it would be possible and potentially straightforward to circumvent such countermeasures by moving PinchBars event tap to kernel space (like other automation tools do), making it invisible to userspace applications like Cubase's frontend. I guess, sooner or later I'll have a closer look into this territorry. But at the time being, I cannot blindly invest more money and more time into this project which I have no financial backing for, yet.

I don't know, maybe more Cubase 13 users will come here trying to find support and we could split at least the financial aspect of this endeavor. Maybe I should link a donation account or something, idk...

</details>

**Comment by p33k** [comment]
> * Is preset "Cubase" still automatically activated when using Cubase 13 (see picture below)?

yes
> * Does the issue persist after changing preset to "None" and then back to "Cubase"?

yes
> * Are vertical and waveform zoom (pinch while pressing CMD / OPT) broken in Cubase 13, too?

no, working. 
CMD + Pinch = Vertical
OPT + Pinch = Waveform
and: CMD + 2 Finger Scroll = Horizontal Zoom
**but just pinch = scrolling strangely up and down**

> * Did any zoom conventions change with Cubase 13? I.e.:
>   4.1. Does normal mouse zoom (not pinch, but scroll while pressing CMD) still work?

CMD + Shift + Mousewheel = Vertical Zoom
CMD + Mousewheel = Horizontal Zoom
>   4.1. Does pressing G/H still work for vertical zoom?

yes: Shift + G/H
>   4.3. Does pressing OPT+G/H still work for waveform zoom?

yes
> * If somebody uses such: Do other automation tools (like Automator, BetterTouchTool, ...) still work with Cubase 13?

have to check that

**Comment by eremsoft** [comment]
Hello!
I will answer your questions a bit later.
They added a new vertical zoom command in 13 and it damaged your app.
https://youtu.be/EILaIuy4gKU?si=btuSb4e3GFHrWs4Y
4.05 min

**Comment by pnoqable** [comment]
Thank you guys! This helps me a lot to understand what's going on here, though - if i recall correctly - they didn't change horizontal mouse zoom, they just added vertical mouse zoom with other modifiers. So I still have to investigate why this broke PinchBar's compatibility. I've contacted a friend who's working with Cubase and Nuendo professionally at his recording studio here in town. Maybe he already upgraded and I can test at his place.. I'll report ASAP.

**Comment by pnoqable** [comment]
Ok, it looks like Cubase 13 handles modifier keys differently than its predecessors. This means this issue can and will be resolved by extending PinchBar's event mapping and uploading a new release. Looks like good news, but there are two major constraints:

1. I won't be able to work on PinchBar for at least a week and having to test with Cubase 13 at my friend's place will take additional time. Also, I don't want to hack this feature into PinchBar and will therefore need some time to rewrite some stuff. So please be patient with me. The reward is reliable software.
2. My membership of Apple's developer program ended a few weaks ago. This is needed to validate and distribute binaries so that macOS acknowledges their code signatures and doesn't throw safety warnings at users hiding buttons to launch them anyways... It is possible, but I don't want to upload unsigned or self-signed binaries. As a result, users will have to build the update by themselves until I renew my membership and upload a new release.

Apple's developer programm costs $100 p.a. which I paid last year to release PinchBar. As I'm quite low on money right now and already invested a lot of effort into PinchBar, I'm putting the renewal of by now. I don't intend to monetize PinchBar, but I want to allocate this cost to its users. [According to GitHub api](https://api.github.com/repos/pnoqable/PinchBar/releases/assets/91794854) there are currently 324 downloads of PinchBar 0.2 and I think a simple call for a donation should get this rolling. I want this donation to be absolutely transparent, so that users can see the progress and when these $100 are reached. But I don't have any experience with this kind of mini-crowd-funding and will therefore have to investigate how this can be achieved. I think this will be my next task.

**Comment by Slaaarti** [comment]
If you put up a crowdfunding link, I'll pay for the Apple dev program.

<edit> By that, I meant I'll pay whatever the program cost is. Other contributions can go towards your time and effort.

**Comment by pnoqable** [comment]
This is very generous of you, Slaaarti! My idea was actually more along the lines of having the costs split by several contributors, so nobody had to spend more then, say, 5 to 10 bucks or so, but the benefit of your offer is clearly to get the update out as soon as possible. I'm currently looking for a free donation-button-solution to put on pinchbar's github page, but I'm not sure which platform to choose. I'm gonna ask a friend who's been working in open source for a long time what he'd recommend for this purpose. Both github sponsors and paypal.me seem to be free but there are some details I'm not so happy with..

Anyway, first thing for me is to get on coding again. As I wrote 2 weeks ago, I didn't have any time to work on PinchBar. But tomorrow, I'm gonna ask my friend whether or when I can come over to his studio for a small (debugging) session. I have some ideas for a first workaround I'd like to test briefly before I start modifying PinchBar's event tap mechanism...

**Comment by pnoqable** [comment]
Good news, everyone!

Now that Cubase 13 is freely available for testing, I was finally able to start working on this issue. By now I know what has changed in Cubase's event handling, and had actually pushed a quickfix a few nights before Christmas. It was just that between all the family visits, I hadn't gotten around to writing this comment yet :)

So if you're somewhat familiar with Xcode, you can check out devel, compile PinchBar yourself and start pinch-to-zoom with Cubase 13!

Please note that this first fix is based on version 0.2 and still requires some work to be compatible with 0.2.8. Therfore, this version doesn't feature middle clicks, Magic Mouse support, one-handed vertical zoom, etc. I'm working on a fully featured update, but want to share this first C13 compatible version with you as soon as possible.

Just take it as an unexpected Christmas present ;)

**Comment by pnoqable** [comment]
Ok nice,

now there's a version that sucessfully combines both, this fix and all the features of 0.2.8. I tested it thoroughly and - unlike its predecessors - it seems very robust, even in the strangest corner case of user input I could come up with. You can test this version by checking out "experimental" (which I linked to this ticket here on GitHub) and compiling PinchBar with Xcode.

It also introduces a new preset ("Cubase 13") which utilizes the new vertical scroll-to-zoom feature of Cubase 13. With this preset, vertical pinch-to-zoom is now aligned with the mouse cursor, just like horizontal pinch-to-zoom already was. With the old one, vertical pinch-to-zoom was mapped to keyboard shortcuts that zoom into/away from the selected track(s), which felt quite odd when triggered via trackpad. Nevertheless, the old preset stays PinchBar's default, as it works with all versions of Cubase whereas the new one requires Cubase 13. If you're only working with Cubase 13 and want vertical pinch-to-zoom to the mouse cursor, just select preset "Cubase 13" in PinchBar's status menu while Cubase is active. It'l remember this preset for Cubase.

Some thoughts about deployment: It might take some time to tidy up this messy branch and merge it back into devel. I'm unsure wheater to put this task off and just create another tag on 0.2.testing. Tough that wouldn't be a real update without an official code signing certificate. I guess, it's getting time to setup a crowdfunding link...

**Comment by justinwelgraven** [comment]
Thank you for your great work! Is there any way of running v0.2.8 without having to compile it ourselves already?

**Comment by Yuli-boop** [comment]
Hi, I jsut discovered this and want to thank you for your amazing work =))
I have been wondering for years why steinberg did not implement this.
I tried it and it works well on Cubase 10.5 Pro on M1. however I recently upgraded to cubase 13 and I dont really understand how to get your “experimental” version to work. Could you maybe quickly explain the steps? thank you so much

**Comment by pnoqable** [comment]
Hi Yuli-boop, thank you for the kind words!

At which point did you get stuck trying to get it to work? If you're already able to build and run it, but did't get its gesture mapping to work: There is one quirk with macOS's security system, when an already installed accessibility app gets a new code signature. I'll try to explain it briefly:

When you downloaded and launched PinchBar for the first time, you had to grant permission in macOS's Accessibility settings (in Privacy & Security) to allow it to map your trackpad gestures to the events that Cubase understands. This permission was then tied to a combination of PinchBar's display name, bundle identifier and its code signature and saved for all later launches. This works as long as this combination doesn't change. It also applies for productive updates of PinchBar as long as they are signed with PinchBars productive code signing certificate.

But if you build PinchBar locally and sign it with a development certificate, macOS's security system won't apply its saved permission for it, as it is now signed with another code signing certificate. Under the hood, macOS refuses to grant permission to this new build, but doesn't show you an alert or something. Even PinchBar itself doesn't get notified by macOS that this permission test failed, so it can't show you a corresponding alert. You just have to find and fix this by yourself. This is an annoyance that always comes up when alternatingly testing local and productive builds of PinchBar.

In such cases, what you always have to do, is to close PinchBar, open macOS's Privacy and Security settings and remove PinchBar completely from Accessibility (by selecting it and clicking on the minus button under this list).
 
The next time you launch any build of PinchBar, as macOS won't find any saved record for its permissions, it'll ask you for them as it did the first time you launched PinchBar. When you grant, macOS will save it tied to the new code signature, fixing your problems for that specific build (and its successors signed with the same certificate).

But, when you later want to use a productive build of PinchBar again, you'll have to redo this steps. Sadly, this can't be automated by PinchBar, you'll always have to do so by yourself.

I hope I have explained everything in a somewhat understandable way. I've been pretty sick since yesterday with the flu, so I can't think very clearly. But don't hesitate to ask again if anything is still unclear.

**Comment by pnoqable** [comment]
Hi@all,
I decided to close this issue, as it is fixed in source. I opened a follow-up with #4 which covers the release of this fix.
@Slaaarti If you're still waiting for this and want to help out, now is your time!

**Comment by pnoqable** [comment]
Hi@all,
I decided to close this issue, as it is fixed in source. I opened a follow-up with #4 which covers its release.
@Slaaarti If you're still waiting for this and want to help out, now is your time!

**Comment by Yuli-boop** [comment]
Hi Lutz, Thanks for your lengthy reply! 
I might have been a bit unclear with my comment. I downloaded the files for your experimental version, but I really don't know anything about coding. All I see are some files in a folder. And somewhere you wrote something about how to get the experimental version to work but I didnt understand any of that. So I thought if the process is easy enough, you could explain it here. But maybe its too complicated =D Also, since you seem to be planning to release 0.3! Really excited for that and I put some money in your crowd fund =)
Best

**Comment by pnoqable** [comment]
> And somewhere you wrote something about how to get the experimental version to work but I didnt understand any of that. So I thought if the process is easy enough, you could explain it here. But maybe its too complicated =D

I'm actually just working on a Youtube video to promote the fundraiser (big thx btw for your contribution!) in which I'm going to include a section about how to build PinchBar from source. Its really not that complicated, but when I continue to explain everything in text, I fear that this all will just get lengthier and lengthier and nobody new to PinchBar will ever take the time to delve into what's actually really easy.

So, that's why now I'm producing promotion videos like I would if I started a new rock group... 🤷‍♂️

**Comment by Yuli-boop** [comment]
If you do, be sure to share the link here =)
